File: app/src/main/java/ru/nehodov/weatherforecast/utils/DailyDiffUtilCallback.java
package ru.nehodov.weatherforecast.utils;

import androidx.recyclerview.widget.DiffUtil;

import java.util.List;

import ru.nehodov.weatherforecast.entities.Daily;

public class DailyDiffUtilCallback extends DiffUtil.Callback {

    private final List<Daily> oldList;
    private final List<Daily> newList;

    public DailyDiffUtilCallback(List<Daily> oldList, List<Daily> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return true;
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return oldList.get(oldItemPosition).getDateTime()
                .equals(newList.get(newItemPosition).getDateTime())
                && oldList.get(oldItemPosition).getWeather()[0].getIcon().equals(
                newList.get(newItemPosition).getWeather()[0].getIcon())
                && oldList.get(oldItemPosition).getTemp()
                .equals(newList.get(newItemPosition).getTemp());
    }
}

File: app/src/main/java/ru/nehodov/weatherforecast/utils/DailyDiffUtilCallback.kt
package ru.nehodov.weatherforecast.utils

import androidx.recyclerview.widget.DiffUtil
import ru.nehodov.weatherforecast.entities.Daily

class DailyDiffUtilCallback(
    private val oldList: List<Daily>,
    private val newList: List<Daily>
) : DiffUtil.Callback() {

    override fun getOldListSize(): Int {
        return oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return true
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldDaily = oldList[oldItemPosition]
        val newDaily = newList.get(newItemPosition)
        return oldDaily.dateTime == newDaily.dateTime &&
                oldDaily.weather[0].icon == newDaily.weather.get(0).icon &&
                oldDaily.temp == newDaily.temp
    }
}